src/sentry/models/releasefile.py

```diff
@@ -236,63 +236,70 @@ class _ArtifactIndexGuard:
     def __init__(self, release: Release, dist: Optional[Distribution]):
         self._release = release
         self._dist = dist
+        self._ident = ReleaseFile.get_ident(ARTIFACT_INDEX_FILENAME, dist and dist.name)
 
     def readable_data(self) -> Optional[dict]:
         """Simple read, no synchronization necessary"""
-        file_ = self._get_file(lock=False)
-        if file_ is not None:
-            with file_.getfile() as fp:
+        try:
+            releasefile = self._releasefile_qs()[0]
+        except IndexError:
+            return None
+        else:
+            with releasefile.file.getfile() as fp:
                 return json.load(fp)
 
     @contextmanager
     def writable_data(self, create: bool) -> ContextManager[_ArtifactIndexData]:
         """Context manager for editable artifact index"""
         with transaction.atomic():
+            created = False
             if create:
-                file_, created = self._get_or_create_file()
+                releasefile, created = self._get_or_create_releasefile()
             else:
-                file_ = self._get_file(lock=True)
-                created = False
-
-            if file_ is None:
+                # Lock the row for editing:
+                # NOTE: Do not select_related('file') here, because we do not
+                # want to lock the File table
+                qs = self._releasefile_qs().select_for_update()
+                try:
+                    releasefile = qs[0]
+                except IndexError:
+                    releasefile = None
+
+            if releasefile is None:
                 index_data = None
```

Comment:
This is the only branch where `index_data` can be `None`, and then the only remaining thing this function does is `yield None`. Can we bail out here instead?

Reply:
You could probably do something like

```
yield None
return
```

but I'm not sure if that would improve readability.